Implement SmartTransaction Line Serialization (MempoolStore Breakdown Part 4)

nopara73 opened the pull request:
I decided to break down PR #2188 to reviewable parts. In this PR I implement SmartTransaction line serialization.

This PR builds upon PR https://github.com/zkSNACKs/WalletWasabi/pull/2201, which is currently hanging in our review hell (waiting for @lontivero's and @molnard's final ACK, and I suspect it won't change anymore, so I was brave enough to create this PR.)

Although we have json serialization implemented for `SmartTransaction`, I decided to use line serialization, because with this I will be able to assert cross process data consistency in a performant manner. (This PR does not contain the below described logic.)

1. We put a system wide async mutex for reading and writing the transaction file, this will ensure concurrency issues.
2. When we want to add a transaction, we just append it to the file. Thus if two processes receive the same transaction, we will append it to the file twice. This results in multiple transactions.
3. When the whole file is first loaded, uniqueness must be ensured: duplicate transactions are to be also removed from the file.
4. When we are to remove a transaction, we'll read the file and remove all the occurrances of that transaction.

I created some drawings to make that logic better understandable:

![image](https://user-images.githubusercontent.com/9156103/64485806-35380f00-d225-11e9-98f5-22ec6b03fb3b.png)

yahiheb commented on WalletWasabi.Tests/UnitTests/SmartTransactionTests.cs:
```diff
@@ -82,46 +82,245 @@ public void FirstSeenBackwardsCompatibility()
+			{
+				yield return new object[] { new SmartTransaction(defaultTx, defaultHeight, blockIndex: blockIndex), defaultNetwork };
+			}
+			foreach (var label in labels)
```
Add separation line.
```suggestion

			foreach (var label in labels)
```

yahiheb commented on WalletWasabi.Tests/UnitTests/SmartTransactionTests.cs:
```diff
@@ -82,46 +82,245 @@ public void FirstSeenBackwardsCompatibility()
+			input = "";
+			Assert.Throws<ArgumentException>(() => SmartTransaction.FromLine(input, network));
+			input = " ";
+			Assert.Throws<ArgumentException>(() => SmartTransaction.FromLine(input, network));
```
Aren't the last two assert repeated?

lontivero replied:
The second one has an space.

yahiheb replied:
No, I meant these two asserts exist already [here](https://github.com/zkSNACKs/WalletWasabi/pull/2216/files#diff-6f2c88d7e4f95ba03f31fc5c62b74431R210) and [here](https://github.com/zkSNACKs/WalletWasabi/pull/2216/files#diff-6f2c88d7e4f95ba03f31fc5c62b74431R212)

nopara73 replied:
Yes, they were duplicated. Removed.

nopara73 commented:
I rebased it to the master now with the merged label PR. It should be easy to review it now.

lontivero commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +125,111 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+
+				return new SmartTransaction(tx, h, bh, bi, sl, ir, fs);
+			}
+			catch
```
Use `catch(Exception e)` instead and log the error because if well it could be a `IndexOutOfRangeException`or `FromatException` it could also be a `NullReferenceException` or similar that means we have a bug in the code.

nopara73 replied:
Makes sense, done!

lontivero commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +125,111 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+			}
+		}
+
+		private static void ParseTransactionHex(Network expectedNetwork, string[] parts, out Transaction tx, out int txHexIndex)
```
Now that we have `tuples` I would prefer to return one instead of `out``parameters.

nopara73 replied:
Done

lontivero commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +125,111 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+			}
+			catch
+			{
+				for (txHexIndex = 0; txHexIndex < parts.Length; txHexIndex++)
```
```suggestion
				for (; txHexIndex < parts.Length; txHexIndex++)
```
Or 
```suggestion
				for (txHexIndex = 1; txHexIndex < parts.Length; txHexIndex++)
```
Then you can remove the `if (txHexeIndex == 1) continue;`

nopara73 replied:
That doesn't work.

lontivero commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +125,111 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+						tx = Transaction.Parse(part, expectedNetwork);
+						break;
+					}
+					catch
```
```suggestion
					catch(FormatException)
```

This prevents hidding possible errors.

nopara73 replied:
We need to catch everything here. We have don't care about possible errors, that just means it's not a transaction, that we don't care about it if it's a null or anything.

nopara73 commented:
> The ParseTransactionHex tries to parse the first element and in case that fails it iterates the rest and tries to parse the other elements but why doing it in this way? Why not just iterate the elements (the first one included) and stop when the transaction is parsed or there is no more elements in parts?

The idea is that we can get a SmartTransaction if we have txhex serialized on any position. Everything else is "just added bonus."

molnard commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +126,115 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+
+				if (!Height.TryParse(heightString, out Height h))
+				{
+					h = Height.Unknown;
```
Use full var name here: height.

nopara73 replied:
Done.

molnard commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +126,115 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+				}
+				if (!uint256.TryParse(blockHashString, out uint256 bh))
+				{
+					bh = null;
```
bh => blockHash

nopara73 replied:
Done.

molnard commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +126,115 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+				}
+				if (!int.TryParse(blockIndexString, out int bi))
+				{
+					bi = 0;
```
use full var name.

nopara73 replied:
Done.

molnard commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +126,115 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+				DateTimeOffset fs = default;
+				if (long.TryParse(firstSeen, out long unixSeconds))
+				{
+					fs = DateTimeOffset.FromUnixTimeSeconds(unixSeconds);
```
use full var name.

nopara73 replied:
Done.

molnard commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +126,115 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+				}
+				if (!bool.TryParse(isReplacementString, out bool ir))
+				{
+					ir = false;
```
use full var name.

nopara73 replied:
Done.

molnard commented on WalletWasabi/Models/SmartTransaction.cs:
```diff
@@ -124,6 +126,115 @@ public static IComparer<SmartTransaction> GetBlockchainComparer()
+			}
+			catch
+			{
+				for (txHexIndex = 0; txHexIndex < parts.Length; txHexIndex++)
```
How is this possible? We have a strict order at ToLine() and also StringSplitOptions.None is used on Split. IMO it is dangerous to let this assumption.

lontivero replied:
Yes, I have the same question. It seems the design assumes that file could be manually modified by someone and write things in a different order and do some other modifications like adding spaces (that's why the `Trim`) and things like that. Why is this?

nopara73 replied:
The idea was so to SmartTransaction can be created from a txhex alone (just like a transaction can be.) But I ruined it for you guys: https://github.com/zkSNACKs/WalletWasabi/pull/2216/commits/d0051ec2b239d9124b7bb85439ac06daecef6dd5

nopara73 commented:
This is not good. 6 days ago I fixed everything, I need this PR to progress further. I addressed all your issues. Merging. @lontivero @molnard Please start every day with github bell notifications, because you're the bottlenecks now.
